File: device-packs/_runtime/sysex-tags.js
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (c) 2026 Matthew Mackes — MAP2 Audio Platform
//
// Shared SysEx name-tag runtime for MPX-1 / IntelFX parser migrations.
// Worklist: T2459-H4

(function (global) {
    'use strict';

    var commonSpecs = [
        ['\\bplate\\b', ['plate', 'reverb']],
        ['\\bhall\\b', ['hall', 'reverb']],
        ['\\broom\\b', ['room', 'reverb']],
        ['\\bspring\\b', ['spring', 'reverb']],
        ['\\bchamber\\b', ['chamber', 'reverb']],
        ['\\becho\\b', ['echo', 'delay']],
        ['\\bdelay\\b', ['delay']],
        ['\\bslap\\b', ['slap', 'delay']],
        ['\\bping\\b', ['pingpong', 'delay']],
        ['\\bchorus\\b', ['chorus']],
        ['\\bpitch\\b', ['pitch']],
        ['\\bshift\\b', ['pitch']],
        ['\\bwhammy\\b', ['pitch']],
        ['\\bdetune\\b', ['pitch']],
        ['\\bharmony\\b', ['pitch']],
        ['\\bocta(ve)?\\b', ['pitch']],
        ['\\beq\\b', ['eq']],
        ['\\bvocal\\b', ['vocal']],
        ['\\bguitar\\b', ['guitar']],
        ['\\bdrum\\b', ['drums']],
        ['\\bsnare\\b', ['drums']],
        ['\\bbass\\b', ['bass']],
        ['\\bambi(ent)?\\b', ['ambient']],
        ['\\bshimmer\\b', ['ambient']],
        ['\\bglass\\b', ['ambient']],
        ['\\bgate\\b', ['gate']],
        ['\\breverse\\b', ['reverse']]
    ];

    var mpx1OnlySpecs = [
        ['\\bflange\\b', ['flange', 'chorus']],
        ['\\bphase\\b', ['phaser', 'chorus']],
        ['\\brotary\\b', ['rotary', 'chorus']],
        ['\\bvibrato\\b', ['vibrato', 'chorus']]
    ];

    var intelfxOnlySpecs = [
        ['\\bhush\\b', ['hush', 'noise_reduction']],
        ['\\bcomp(ressor)?\\b', ['compressor']],
        ['\\bwah\\b', ['wah']],
        ['\\bflange\\b', ['flanger', 'chorus']],
        ['\\bphase\\b', ['phaser']],
        ['\\brotary\\b', ['rotary', 'chorus']],
        ['\\btremolo\\b', ['tremolo']],
        ['\\bvibrato\\b', ['vibrato', 'chorus']],
        ['\\breverb\\b', ['reverb']],
        ['\\bclean\\b', ['clean']],
        ['\\bcrunch\\b', ['crunch']],
        ['\\blead\\b', ['lead']],
        ['\\bacoustic\\b', ['acoustic']]
    ];

    function compileTagRules(specs) {
        return specs.map(function (item) {
            return { pattern: new RegExp(item[0], 'i'), tags: item[1].slice() };
        });
    }

    function autoTagFromName(name, rules) {
        var out = [];
        var seen = Object.create(null);
        for (var i = 0; i < rules.length; ++i) {
            var rule = rules[i];
            if (!rule.pattern.test(name)) continue;
            for (var j = 0; j < rule.tags.length; ++j) {
                var tag = rule.tags[j];
                if (seen[tag]) continue;
                seen[tag] = true;
                out.push(tag);
            }
        }
        return out;
    }

    function buildMpx1TagRules() {
        return compileTagRules(commonSpecs.concat(mpx1OnlySpecs));
    }

    function buildIntelfxTagRules() {
        return compileTagRules(commonSpecs.concat(intelfxOnlySpecs));
    }

    global.MAP2SysexTags = {
        autoTagFromName: autoTagFromName,
        buildMpx1TagRules: buildMpx1TagRules,
        buildIntelfxTagRules: buildIntelfxTagRules
    };
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: app/services/sysex_tags.py
"""Shared SysEx program-name auto-tagging utilities.

Worklist: T2459-H4 (SysEx parser consolidation slice)
"""

from __future__ import annotations

import re
from typing import Iterable


TagMap = list[tuple[re.Pattern[str], list[str]]]


_COMMON_TAG_SPECS: tuple[tuple[str, tuple[str, ...]], ...] = (
    (r"\bplate\b", ("plate", "reverb")),
    (r"\bhall\b", ("hall", "reverb")),
    (r"\broom\b", ("room", "reverb")),
    (r"\bspring\b", ("spring", "reverb")),
    (r"\bchamber\b", ("chamber", "reverb")),
    (r"\becho\b", ("echo", "delay")),
    (r"\bdelay\b", ("delay",)),
    (r"\bslap\b", ("slap", "delay")),
    (r"\bping\b", ("pingpong", "delay")),
    (r"\bchorus\b", ("chorus",)),
    (r"\bpitch\b", ("pitch",)),
    (r"\bshift\b", ("pitch",)),
    (r"\bwhammy\b", ("pitch",)),
    (r"\bdetune\b", ("pitch",)),
    (r"\bharmony\b", ("pitch",)),
    (r"\bocta(ve)?\b", ("pitch",)),
    (r"\beq\b", ("eq",)),
    (r"\bvocal\b", ("vocal",)),
    (r"\bguitar\b", ("guitar",)),
    (r"\bdrum\b", ("drums",)),
    (r"\bsnare\b", ("drums",)),
    (r"\bbass\b", ("bass",)),
    (r"\bambi(ent)?\b", ("ambient",)),
    (r"\bshimmer\b", ("ambient",)),
    (r"\bglass\b", ("ambient",)),
    (r"\bgate\b", ("gate",)),
    (r"\breverse\b", ("reverse",)),
)

_MPX1_ONLY_TAG_SPECS: tuple[tuple[str, tuple[str, ...]], ...] = (
    (r"\bflange\b", ("flange", "chorus")),
    (r"\bphase\b", ("phaser", "chorus")),
    (r"\brotary\b", ("rotary", "chorus")),
    (r"\bvibrato\b", ("vibrato", "chorus")),
)

_INTELFX_ONLY_TAG_SPECS: tuple[tuple[str, tuple[str, ...]], ...] = (
    (r"\bhush\b", ("hush", "noise_reduction")),
    (r"\bcomp(ressor)?\b", ("compressor",)),
    (r"\bwah\b", ("wah",)),
    (r"\bflange\b", ("flanger", "chorus")),
    (r"\bphase\b", ("phaser",)),
    (r"\brotary\b", ("rotary", "chorus")),
    (r"\btremolo\b", ("tremolo",)),
    (r"\bvibrato\b", ("vibrato", "chorus")),
    (r"\breverb\b", ("reverb",)),
    (r"\bclean\b", ("clean",)),
    (r"\bcrunch\b", ("crunch",)),
    (r"\blead\b", ("lead",)),
    (r"\bacoustic\b", ("acoustic",)),
)


def _compile_specs(specs: Iterable[tuple[str, tuple[str, ...]]]) -> TagMap:
    return [(re.compile(pattern, re.I), list(tags)) for pattern, tags in specs]


def compile_mpx1_tag_map() -> TagMap:
    return _compile_specs((*_COMMON_TAG_SPECS, *_MPX1_ONLY_TAG_SPECS))


def compile_intelfx_tag_map() -> TagMap:
    return _compile_specs((*_COMMON_TAG_SPECS, *_INTELFX_ONLY_TAG_SPECS))


def auto_tag_from_name(name: str, tag_map: TagMap) -> list[str]:
    tags: list[str] = []
    seen: set[str] = set()
    for pattern, tag_list in tag_map:
        if pattern.search(name):
            for tag in tag_list:
                if tag not in seen:
                    tags.append(tag)
                    seen.add(tag)
    return tags

